worker/dblogpruner/worker.go

```diff
@@ -17,80 +18,82 @@ import (
-
-// NewLogPruneParams returns a LogPruneParams initialised with default values.
-func NewLogPruneParams() *LogPruneParams {
-	return &LogPruneParams{
-		PruneInterval: DefaultPruneInterval,
+func (config Config) Validate() error {
+	if config.State == nil {
+		return errors.NotValidf("nil State")
+	}
+	if config.Clock == nil {
+		return errors.NotValidf("nil Clock")
 	}
+	if config.PruneInterval <= 0 {
+		return errors.NotValidf("non-positive PruneInterval")
+	}
+	return nil
 }
 
-// New returns a worker which periodically wakes up to remove old log
-// entries stored in MongoDB. This worker is intended to run just
-// once, on the MongoDB master.
-func New(st *state.State, params *LogPruneParams) worker.Worker {
-	w := &pruneWorker{
-		st:     st,
-		params: params,
+// NewWorker returns a worker which periodically wakes up to remove old log
+// entries stored in MongoDB. This worker must not be run in more than one
+// agent concurrently.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
 	}
-	return jworker.NewSimpleWorker(w.loop)
+	w := &pruneWorker{config: config}
+	return jworker.NewSimpleWorker(w.loop), nil
 }
 
 type pruneWorker struct {
-	st     *state.State
-	params *LogPruneParams
+	config Config
 }
 
 func (w *pruneWorker) loop(stopCh <-chan struct{}) error {
-
-	controllerConfigWatcher := w.st.WatchControllerConfig()
+	controllerConfigWatcher := w.config.State.WatchControllerConfig()
 	defer worker.Stop(controllerConfigWatcher)
 
 	var (
 		maxLogAge               time.Duration
 		maxCollectionMB         int
 		controllerConfigChanges = controllerConfigWatcher.Changes()
-		// We will also get an initial event, but need to ensure that event is
-		// received before doing any pruning.
-		haveConfig = false
+		pruneTimer              clock.Timer
+		pruneCh                 <-chan time.Time
 	)
-	p := w.params
 
 	for {
 		select {
 		case <-stopCh:
 			return tomb.ErrDying
+
 		case _, ok := <-controllerConfigChanges:
 			if !ok {
 				return errors.New("controller configuration watcher closed")
 			}
-			controllerConfig, err := w.st.ControllerConfig()
+			controllerConfig, err := w.config.State.ControllerConfig()
 			if err != nil {
 				return errors.Annotate(err, "cannot load controller configuration")
 			}
-			haveConfig = true
 			newMaxAge := controllerConfig.MaxLogsAge()
 			newMaxCollectionMB := controllerConfig.MaxLogSizeMB()
 			if newMaxAge != maxLogAge || newMaxCollectionMB != maxCollectionMB {
 				logger.Infof("log pruning config: max age: %v, max collection size %dM", newMaxAge, newMaxCollectionMB)
 				maxLogAge = newMaxAge
 				maxCollectionMB = newMaxCollectionMB
 			}
-			continue
-		case <-time.After(p.PruneInterval):
-			if !haveConfig {
-				continue
+			if pruneTimer == nil {
+				pruneTimer = w.config.Clock.NewTimer(w.config.PruneInterval)
+				pruneCh = pruneTimer.Chan()
+				defer pruneTimer.Stop()
```

Comment:
is there a reason to defer creating the pruneTimer until now?
Could we just set up the pruneTimer outside of the loop?
Also, shouldn't we be calling pruneTimer.Reset(w.config.PruneInterval) when the config changes?

Reply:
> is there a reason to defer creating the pruneTimer until now?

Yes, the reason is that we don't have any configuration until we get the first watcher notification. I'll comment.

> Also, shouldn't we be calling pruneTimer.Reset(w.config.PruneInterval) when the config changes?

Nope, pruning should be every <configured interval>, from when we get the first config notification. The interval is statically configured via the manifold config.